addons/sale/sale.py

```diff
@@ -588,6 +589,15 @@ def action_cancel(self, cr, uid, ids, context=None):
                         _('Cannot cancel this sales order!'),
                         _('First cancel all invoices attached to this sales order.'))
                 inv.signal_workflow('invoice_cancel')
+
+            # Cancel all procurements linked to the sale order
+            for orderline in sale.order_line:
+                procurement_obj.cancel(cr, uid,
+                    procurement_obj.search(cr, uid, [
+                        ('group_id', '=', orderline.procurement_ids.group_id.id)
+                        ], context=context),
```

Comment:
If you have the procurement_ids from the orderline, you should not be searching for the ones in stock,...  As indicated in Github also, if you cancel the procurement in Customers, it will automatically cancel the once before (in case of propagate cancel/split)

It will also still give an error in sale_stock where the system still tries to avoid you cancel a sale order with a delivery order.